### PR TITLE
SearchKit - New getLinks API action for advanced link handling

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
This adds a new generic APIv4 action: 'getLinks', and uses it in SearchKit to improve link handling.

- Unblocks #27717 
- Fixes [dev/core#4804 SearchKit: New toolbar in SearchKit doesn't show for contacts entity](https://lab.civicrm.org/dev/core/-/issues/4804)
- Unblocks [dev/core#4674 civicrm_admin_ui blocks links from oauth-client](https://lab.civicrm.org/dev/core/-/issues/4674)
- Unblocks https://github.com/civicrm/civicrm-core/pull/28810

Before
----------------------------------------
SearchKit can display simple links, but some entities require more complex logic.
In the metadata, links for each entity are in a flat static array in the DAO, with some logic sprinkled around `SearchKit` and `CRM_Core_Action` to transform that into something useful.

After
----------------------------------------
The new API action can check permissions, replace tokens, and output translated labels and icons. The hookability of the API give additional flexibility for entities that need extra logic.

This enables SearchKit to work with the complex logic in e.g. Activity or Contact links.

image

Technical Details
----------------------------------------
This centralizes the link rendering done by SearchKit, and will allow SearchKit to handle more dynamic stuff like multiple "Add" links in the toolbar.

### CODE REVIEW

**Comment by mattwire** [comment]
@colemanw Would this work with the "toolbar" button links as well?

**Comment by colemanw** [comment]
@mattwire yes that was the biggest motivation behind this PR, to allow entities with > 1 "create" link (Activities, ECK, MembershipStatus) to be able to implement their special logic for dynamically-generated add links which SK will render in the toolbar as a dropdown instead of simple button.
Did you have another one in mind?

**Comment by colemanw** [comment]
@mattwire the next step is to adapt SearchKit to take advantage of this new API... and that's turning into a refactoring nightmare. But I think this PR is at a good mergeable point now. It works and has working tests and it gives the output we need for the entities with complex links.

**Comment by colemanw** [comment]
Update: This PR now includes the full-stack incorporating the new API action into SearchKit (see updated description with screenshot).

File: tests/phpunit/api/v4/Action/GetLinksTest.php [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+      ->setExpandMultiple(TRUE)
+      ->execute();
+    $this->assertStringContainsString('ct=Individual', $links[0]['path']);
+    $this->assertEquals('Add Individual', $links[0]['text']);
+    $this->assertEquals(0, $links[0]['weight']);
+    $this->assertNull($links[0]['api_values']);
+    $links = $links->indexBy('text');
+    $this->assertEquals('fa-squirrel', $links['Add Squirrel']['icon']);
+    $this->assertGreaterThan($links['Add Rabbit']['weight'], $links['Add Squirrel']['weight']);
+    $this->assertGreaterThan($links['Add Chipmunk']['weight'], $links['Add Rabbit']['weight']);
```
Comment (eileenmcnaughton) [inline tests/phpunit/api/v4/Action/GetLinksTest.php:81]: Chipmunks are not a valid test attribute

**Comment by eileenmcnaughton** [comment]
The code here is mostly new or moved so despite this looking like a lot of change it's impact on existing code is minor & test cover is added

**Pull request merged**